src/lib/arch/riscv/Architecture.cc

```diff
@@ -184,6 +198,17 @@ uint8_t Architecture::predecode(const void* ptr, uint8_t bytesAvailable,
     bool success =
         cs_disasm_iter(capstoneHandle, &encoding, &size, &address, &rawInsn);
 
+    if (!success || rawInsn.opcode == RISCV_INS_INVALID) {
+      // Invalid decoding, potentially read over the end of the valid buffer.
+      // BAIL. Do not add to cache as may be incorrect data
+      return 0;
+    }
```

Comment:
Yes, however, this clause may trigger in the following scenario: the buffer (2*fetchBlockSize_) gets filled with data, fetch unit continues predecoding through the buffer, now only half of the buffer contains valid data - the other half dirty from previous use. Due to the use of compressed instructions there is no 4 byte alignment (2 byte instead). With fetch block sizes which are a multiple of 4, half of a 32 bit instruction may be fetched into the end of the valid portion of the buffer. When reading 4 bytes from here, the first half will be from the valid section and the second half from the invalid. This often creates and invalid instruction which is caught here. We do not want to cache this (or carry on in any manner) as it was incorrect input data that caused this.

It may be possible that the concatenation of valid and invalid data could result in a valid instruction encoding being generated. This will then be decoded by capstone as normal, however the next clause should catch this and prevent this from propagating further.

It may be possible to prevent this entirely by modifying the memcpy to not copy over the end of the valid data actually (by checking the bytes available). Would this be a preferable solution? It would incur extra minor per tick logic as compared to the current which is major per instruction logic.

Reply:
Implementation changed so this clause is invalid. Resolving